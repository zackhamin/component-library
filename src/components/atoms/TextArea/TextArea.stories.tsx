import { Story } from '@storybook/react';
import { TextArea, TextAreaCombinedProps } from './TextArea';

export default {
	title: 'atoms/TextArea',
	component: TextArea,
};

const Template: Story<TextAreaCombinedProps> = (args) => <TextArea {...args} />;

const defaultProps = {
	id: 'inputID',
	name: 'inputName',
	placeholder: 'Placeholder text',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Disabled = Template.bind({});
Disabled.args = {
	...defaultProps,
	isDisabled: true,
	placeholder: '',
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
