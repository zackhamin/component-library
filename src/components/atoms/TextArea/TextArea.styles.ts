import { hexToRGB } from 'Utils/utils';
import styled from 'styled-components';
import { colors, typography } from 'Styles';

const height = 4.8;

const StyledTextArea = styled.textarea`
	min-height: ${height}rem;
	resize: vertical;
	width: 100%;
	padding: 1.6rem;
	font-size: ${typography.size.s3}rem;
	font-family: ${typography.type.body};
	color: ${colors.offBlack};
	border: 1px solid ${colors.offBlack};
	border-radius: 0.4rem;

	&::placeholder {
		color: ${colors.grey500};
	}

	&:disabled {
		background-color: ${colors.lightGrey};
		border-color: ${colors.grey500};
	}

	&:focus-visible {
		outline: none;
		border-color: ${colors.emerald};
		box-shadow: 0 0 6px ${hexToRGB(colors.emerald, 0.4)};
	}

	&.is-invalid {
		border-color: ${colors.error};
		// Compensation for icon
		padding-right: 5rem;
	}

	&.is-success {
		// Compensation for icon
		padding-right: 5rem;
	}
`;

const StyledTextAreaWrapper = styled.div`
	position: relative;
`;

const StyledTextAreaIcon = styled.div`
	position: absolute;
	top: ${height / 2}rem;
	right: 1.5rem;
	transform: translateY(-50%);
`;

export { StyledTextArea, StyledTextAreaWrapper, StyledTextAreaIcon };
