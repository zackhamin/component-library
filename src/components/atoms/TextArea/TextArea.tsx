import { colors } from 'Styles';
import { useInputContext } from 'Hooks/useInputContext';
import { BaseInputProps, ConditionalFormProps, Wrap } from '@Types/form';
import { ChangeEvent, forwardRef } from 'react';
import { Check, Error } from 'Atoms/Icons';
import {
	StyledTextArea,
	StyledTextAreaIcon,
	StyledTextAreaWrapper,
} from './TextArea.styles';

export interface TextAreaProps extends BaseInputProps<HTMLTextAreaElement> {
	placeholder?: string;
	value?: string;
	onChange?: (e: ChangeEvent<HTMLTextAreaElement>) => void;
	wrap?: Wrap;
	rows?: number;
}

export type TextAreaCombinedProps = TextAreaProps & ConditionalFormProps;

const TextArea = forwardRef<HTMLTextAreaElement, TextAreaCombinedProps>(
	(
		{
			name,
			id,
			isDisabled,
			isInvalid,
			isRequired,
			isSuccess,
			placeholder,
			className,
			value,
			onChange,
			...rest
		},
		ref
	) => {
		const { inputClasses, ctx } = useInputContext({
			id,
			name,
			isInvalid,
			isRequired,
			isSuccess,
			className,
		});

		return (
			<StyledTextAreaWrapper>
				<StyledTextArea
					className={inputClasses}
					name={ctx.name}
					id={ctx.id}
					ref={ref}
					placeholder={placeholder}
					aria-invalid={ctx.isInvalid}
					aria-describedby={ctx.isInvalid ? `${ctx.id}--feedback` : undefined}
					required={ctx.isRequired}
					disabled={isDisabled}
					value={value}
					onChange={onChange}
					{...rest}
				/>

				{ctx.isInvalid && !isDisabled && (
					<StyledTextAreaIcon>
						<Error fontSize="1.8rem" color={colors.error} data-testid="ERROR" />
					</StyledTextAreaIcon>
				)}

				{ctx.isSuccess && !isDisabled && (
					<StyledTextAreaIcon>
						<Check
							fontSize="1.8rem"
							color={colors.emerald75}
							data-testid="CHECK"
						/>
					</StyledTextAreaIcon>
				)}
			</StyledTextAreaWrapper>
		);
	}
);

export { TextArea };
