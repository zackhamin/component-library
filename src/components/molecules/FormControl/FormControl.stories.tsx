import { FormError } from 'Atoms/FormError';
import { FormLabel } from 'Atoms/FormLabel';
import { Story } from '@storybook/react';
import { TextInput } from 'Atoms/TextInput';
import { FormControl, FormControlProps } from './FormControl';

export default {
	title: 'molecules/FormControl',
	component: FormControl,
	argTypes: {
		isSuccess: {
			control: {
				type: 'boolean',
			},
		},
		isInvalid: {
			control: {
				type: 'boolean',
			},
		},
	},
};

const Template: Story<FormControlProps> = (args) => (
	<FormControl {...args}>
		<FormLabel>Label</FormLabel>
		<TextInput />
		<FormError>Error Message</FormError>
	</FormControl>
);
const defaultProps = {
	name: 'testName',
	id: 'testId',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Required = Template.bind({});
Required.args = {
	...defaultProps,
	isRequired: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
