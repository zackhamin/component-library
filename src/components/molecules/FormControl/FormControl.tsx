import { ConditionalFormProps } from '@Types/form';
import { createContext } from 'react';

interface FormControlContext {
	name?: string;
	id: string;
	isRequired?: boolean;
	isInvalid?: boolean;
	isSuccess?: boolean;
}

interface BaseFormControlProps extends FormControlContext {
	children: React.ReactNode;
	className?: string;
}

export type FormControlProps = BaseFormControlProps & ConditionalFormProps;

export const FormControlContext = createContext<FormControlContext | undefined>(
	undefined
);

const FormControl = ({
	children,
	name,
	id,
	isInvalid,
	isRequired,
	isSuccess,
	className,
}: FormControlProps) => {
	return (
		<FormControlContext.Provider
			value={{ name, id, isRequired, isInvalid, isSuccess }}
		>
			<div className={className}>{children}</div>
		</FormControlContext.Provider>
	);
};

export { FormControl };
