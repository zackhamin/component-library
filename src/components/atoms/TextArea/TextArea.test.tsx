import { FormControl } from 'Molecules/FormControl';
import { TextArea, TextAreaCombinedProps } from '../TextArea';
import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps: TextAreaCombinedProps = {
	id: 'textareaID',
	name: 'textareaName',
	placeholder: 'Placeholder text',
	className: 'additional-class',
};

describe('TextArea', () => {
	it('renders correctly', () => {
		const { container } = render(<TextArea {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});

	it('adds ID, class, type & name to element', () => {
		render(
			<TextArea
				{...defaultProps}
				rows={10}
				minLength={10}
				maxLength={100}
				wrap={'hard'}
			/>
		);

		const textarea = screen.getByRole('textbox');

		expect(textarea).toHaveAttribute('id', 'textareaID');
		expect(textarea).toHaveAttribute('name', 'textareaName');
		expect(textarea).toHaveClass('additional-class');

		expect(textarea).toHaveAttribute('rows', '10');
		expect(textarea).toHaveAttribute('minLength', '10');
		expect(textarea).toHaveAttribute('maxLength', '100');
		expect(textarea).toHaveAttribute('wrap', 'hard');

		expect(textarea).not.toHaveAttribute('type');
		expect(textarea).not.toHaveAttribute('aria-invalid');
		expect(textarea).not.toHaveAttribute('aria-describedby');
	});
});

describe('Textarea has disabled state', () => {
	it('adds disabled attribute to element', () => {
		render(<TextArea {...defaultProps} isDisabled />);
		expect(screen.getByRole('textbox')).toHaveAttribute('disabled');
	});
});

describe('Textarea has invalid state', () => {
	it('renders with invalid class', () => {
		render(<TextArea {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveClass('is-invalid');
	});

	it('adds aria-invalid attribute to element', () => {
		render(<TextArea {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveAttribute('aria-invalid', 'true');
	});

	it('should have aria-describedby attribute based on id', () => {
		render(<TextArea {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveAttribute(
			'aria-describedby',
			'textareaID--feedback'
		);
	});

	it('should display an SVG error icon', () => {
		render(<TextArea {...defaultProps} isInvalid />);
		expect(screen.getAllByTestId('ERROR')).toHaveLength(1);
	});
});

describe('Textarea has success state', () => {
	it('renders with success class', () => {
		render(<TextArea {...defaultProps} isSuccess />);
		expect(screen.getByRole('textbox')).toHaveClass('is-success');
	});

	it('does not have aria-invalid & aria-describedby attributes', () => {
		render(<TextArea {...defaultProps} isSuccess />);
		expect(screen.getByRole('textbox')).not.toHaveAttribute('aria-invalid');
		expect(screen.getByRole('textbox')).not.toHaveAttribute('aria-describedby');
	});

	it('should display an SVG error icon', () => {
		render(<TextArea {...defaultProps} isSuccess />);
		expect(screen.getAllByTestId('CHECK')).toHaveLength(1);
	});
});

describe('Textarea change handler', () => {
	it('should call onChange on user input', () => {
		const onChangeHandler = jest.fn();
		const inputValue = 'input value';

		render(<TextArea {...defaultProps} onChange={onChangeHandler} />);

		const input = screen.getByRole('textbox');

		fireEvent.change(input, { target: { value: inputValue } });

		expect(onChangeHandler).toHaveBeenCalled();
	});
});

describe('TextArea wrapped in FormControl', () => {
	it('should have the correct ID passed from Form Control', () => {
		render(
			<FormControl id="control">
				<TextArea />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('id', 'control');
	});

	it('should have required attr when isRequired is passed from Form Control', () => {
		render(
			<FormControl id="control" isRequired>
				<TextArea />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('required');
	});

	it('should have invalid attr and class when isInvalid passed from Form Control', () => {
		render(
			<FormControl id="control" isInvalid>
				<TextArea />
			</FormControl>
		);

		const textarea = screen.getByRole('textbox');

		expect(textarea).toHaveAttribute('aria-invalid', 'true');
		expect(textarea).toHaveClass('is-invalid');
	});

	it('should have correct class when isSuccess passed from Form Control', () => {
		render(
			<FormControl id="control" isSuccess>
				<TextArea />
			</FormControl>
		);

		const textarea = screen.getByRole('textbox');

		expect(textarea).toHaveClass('is-success');
	});
});
