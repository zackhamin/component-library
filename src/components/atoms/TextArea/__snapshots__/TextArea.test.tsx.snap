// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
.c1 {
  min-height: 4.8rem;
  resize: vertical;
  width: 100%;
  padding: 1.6rem;
  font-size: 1.6rem;
  font-family: Helvetica,Arial,sans-serif;
  color: #333333;
  border: 1px solid #333333;
  border-radius: 0.4rem;
}

.c1::-webkit-input-placeholder {
  color: #979797;
}

.c1::-moz-placeholder {
  color: #979797;
}

.c1:-ms-input-placeholder {
  color: #979797;
}

.c1::placeholder {
  color: #979797;
}

.c1:disabled {
  background-color: #EEEEEF;
  border-color: #979797;
}

.c1:focus-visible {
  outline: none;
  border-color: #00594C;
  box-shadow: 0 0 6px rgba(0,89,76,0.4);
}

.c1.is-invalid {
  border-color: #D61616;
  padding-right: 5rem;
}

.c1.is-success {
  padding-right: 5rem;
}

.c0 {
  position: relative;
}

<div>
  <div
    class="c0"
  >
    <textarea
      class="c1 additional-class"
      id="textareaID"
      name="textareaName"
      placeholder="Placeholder text"
    />
  </div>
</div>
`;
